### ISSUE TITLE
Encode parquet null count statistics correctly

### ISSUE THREAD

**Issue by alamb** [apache/arrow-rs#6256]
**Describe the bug**
While reviewing https://github.com/apache/arrow-rs/pull/6216 from @Michael-J-Ward (🙏 ) I am pretty sure the arrow-rs parquet writer does not do the correct thing with respect to null statistics

Specifically, when there are no nulls in the data, the writer does not emit a value for null_count in the thrift metadata (it writes the equivalent of `None`) -- it should instead write the equivalent of `Some(0)`

This will not cause issues for people using parquet-rs to read and write data as the reader also (incorrectly) reports `Some(0)` when the thrift metadata has `None`

**To Reproduce**
TBD

**Expected behavior**
* When writing statistics for data without nulls, the parquet-rs writer should write `Some(0)`
* When reading statistics, the parquet-rs reader should read `None`

**Additional context**

**Comment by mapleFU** [comment]
> When reading statistics, the parquet-rs reader should read `None`

I'm a bit curious here, since like in page-index ( https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L1088 ) and page header v2 ( https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L633 ), the null-count is better to have. Should it also `Some(0)` to represent "nullable type has null count, and it's 0" ?

**Comment by mapleFU** [comment]
> I think the parquet-rs writer always has the null count when writing statistics so it will always write Some(...)

I've checked parquet-java and parquet-c++, all of these seems to write null-count with 0, but I don't know the case of legacy data. Maybe @wgtmac knows?

**Comment by alamb** [comment]
I agree it is pretty unlikely to find one of these files in practice

**Comment by mapleFU** [comment]
https://github.com/apache/parquet-java/blob/d4384d3f2e7703fab6363fd9cd80a001e9561db2/parquet-hadoop/src/main/java/org/apache/parquet/filter2/statisticslevel/StatisticsFilter.java#L93

LOL, seems parquet-java uses None == 0? Maybe I can draft a pr in parquet-format to enforcing this or get more info about this...🤔

**Comment by alamb** [comment]
> seems parquet-java distinguishes None and 0. Maybe I can draft a pr in parquet-format to enforcing this or get more info about this...🤔

I am not sure how important it is to enforce that this statistics field is set

Note that all the parquet files written by the rust writer so far will have had their null counts set to `None`...

**Comment by mapleFU** [comment]
From my aspects, 0 and None is different, and parquet-c++ / parquet-java always writes the null_count if has any value. Perhaps there're some old file write it to 0, and make filter pruning becomes buggy, lol

I'll draft a discuss in mailling list about this tomorrow, a bit late in utc-8 and I have to sleep :-)

**Comment by alamb** [comment]
> I'll draft a discuss in mailling list about this tomorrow, a bit late in utc-8 and I have to sleep :-)

Thank you . Pleasant dreams 💤

**Comment by alamb** [comment]
> From my aspects, 0 and None is different, 

Yes, this makes sense to me

> and parquet-c++ / parquet-java always writes the null_count if has any value.

Yes, this is what I think the rust writer should do too (I am making a PR to make it so)

**Comment by alamb** [comment]
I made a PR with a proposed fix: https://github.com/apache/arrow-rs/pull/6257 which turned out to be larger than I expected (though it is mostly tests)

**Comment by Michael-J-Ward** [comment]
> page header v2 ( https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L633 ), the null-count is better to have

Is there a semantic difference between `DataPageHeaderV2.num_nulls` and `DataPageHeaderV2.statistics.null_count`?

`parquet-rs` currently uses the same value for both, and truncates `num_nulls`.

**`DataPageHeaderV2.num_nulls`**
https://github.com/apache/arrow-rs/blob/0f7116bf81a4cce10a6e3ae4ed9b6ee01a20c462/parquet/src/column/writer/mod.rs#L994-L1004

**`DataPageHeaderV2.statistics.null_count`**
https://github.com/apache/arrow-rs/blob/0f7116bf81a4cce10a6e3ae4ed9b6ee01a20c462/parquet/src/column/writer/mod.rs#L889C1-L906C11

---

This stuck out to me because #6257 adds the checked conversion for `u64 -> i64`.

**Comment by alamb** [comment]
I am not sure to be honest 🤔

**Comment by mapleFU** [comment]
> Is there a semantic difference between DataPageHeaderV2.num_nulls and DataPageHeaderV2.statistics.null_count?

I think no 🤔

Besides the DataPage V2 seems not used by most community. I can also mention this is maillist.

**Comment by mapleFU** [comment]
I've update it here: https://lists.apache.org/thread/oqd9lt7p0jtf217hg0667xhk0zvwbgvt
